src/importer.rs

```diff
@@ -0,0 +1,268 @@
+use std::collections::hash_map::Entry;
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::error;
+use std::fmt;
+use std::fs;
+use std::io::{self, Read};
+
+use crc32fast;
+use tobj;
+
+use crate::viewport_renderer::{Index, Vertex};
+
+#[derive(Debug, PartialEq)]
+pub enum ImporterError {
+    FileNotFound,
+    PermissionDenied,
+    InvalidStructure,
+    Other,
+}
+
+impl fmt::Display for ImporterError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            ImporterError::FileNotFound => write!(f, "File was not found."),
+            ImporterError::InvalidStructure => write!(f, "The obj file is not valid."),
+            ImporterError::PermissionDenied => write!(f, "Permission denied."),
+            ImporterError::Other => write!(f, "Unexpected error happened."),
+        }
+    }
+}
+
+impl error::Error for ImporterError {}
+
+impl From<io::Error> for ImporterError {
+    fn from(err: io::Error) -> Self {
+        match err.kind() {
+            io::ErrorKind::NotFound => ImporterError::FileNotFound,
+            io::ErrorKind::PermissionDenied => ImporterError::PermissionDenied,
+            _ => ImporterError::Other,
+        }
+    }
+}
+
+impl From<tobj::LoadError> for ImporterError {
+    fn from(_err: tobj::LoadError) -> Self {
+        ImporterError::InvalidStructure
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Model {
+    pub name: String,
+    pub vertices: Vec<Vertex>,
+    pub indices: Vec<Index>,
+}
+
+#[derive(Debug)]
+pub struct FileMetadata {
+    checksum: u32,
+    last_modified: std::time::SystemTime,
+}
+
+/// `Importer` takes care of importing of obj files and caching of their
+/// internal representations. It holds paths to files, their metadata
+/// (checksums, timestamps) and models of parsed obj files.
+#[derive(Debug, Default)]
+pub struct Importer {
+    path_metadata: HashMap<String, FileMetadata>,
+    loaded_models: HashMap<u32, Vec<Model>>,
+}
+
+impl Importer {
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Tries to import obj file from given `path`. If file was already imported
+    /// and its timestamp is identical, parsed models are returned from cache.
+    /// Otherwise, file is read, checksum calculated and cache is checked whether
+    /// given file contents were already saved. If not, obj file is parsed and
+    /// cached.
+    pub fn import_obj(&mut self, path: &str) -> Result<Vec<Model>, ImporterError> {
+        let mut file = fs::File::open(path)?;
+        let file_modified = file
+            .metadata()
+            .and_then(|metadata| metadata.modified())
+            .expect("obj file should return its modified timestamp");
```

Comment:
Any particular reason why you didn't just `?` the error? You already have the `From` impl for it. Unless I am missing something, the code would be both shorter and more resilient.